Support iframes (nested browsing contexts) in selection event handling

acusti opened the pull request:
I tried to rebase #7936 with latest master and everything was thoroughly screwed up. So rather than trying to merge master, I just made a new clean branch off master and cherry-picked the commits over. I’m not closing the original PR at this point, however, because there seem to be many interested people following the activity over there, and I don’t want to prevent them from doing so. I’ll make new built versions of everything off of these commits applied to 15-stable and update the npm / yarn instructions on using those when I get a chance.

I’ve also made a commit with a couple guards to accommodate varying cross-browser behaviors. The guards are for https://github.com/facebook/react/pull/7936#issuecomment-270568409 and https://github.com/summernote/summernote/issues/1057

acusti commented:
Just fixed an issue that resulted from me incompletely recreating the changes of #7936 when I originally created this new PR. Then I rebased with latest master.

The CircleCI build has 5 React DOM test failures:

```
 FAIL  src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js (38.268s)
 FAIL  src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js (6.354s)
...
 FAIL  src/renderers/dom/shared/__tests__/ReactMount-test.js
...
 FAIL  src/renderers/dom/shared/__tests__/ReactRenderDocument-test.js
...
 FAIL  src/renderers/dom/shared/__tests__/ReactDOMTextComponent-test.js
```

First failure has a lot of output related to `console.log` along the lines of:
```
    console.log src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js:45
      We expected 2 warning(s), but saw 1 warning(s).
```

Others have no info in the test output. I’m at a loss as to what’s causing the failures in CI. I ran the tests locally and each one of those tests listed above passed. @spicyj Any suggestions or intuitions about the cause?

nhunzaker commented:
> I tried to rebase #7936 with latest master and everything was thoroughly screwed up. 

@acusti Haha, I know that feeling... **Thank you so much for doing all of this work and sticking with us.**

I'm working through this PR now. In the mean time: I had a couple of questions:

1. Is it safe to close out https://github.com/facebook/react/pull/7866 and https://github.com/facebook/react/pull/7936?
2. CI is only failing because of some formatting. Could you run `yarn prettier` on this branch and send up the result?

On my end, I need to follow up on the conversation in the prior PRs, some of the associated issues linked on projects, and better understand how React input selection restoration works.

I promise to respond in no later than 2 days, even if it's just to say where I'm at.

acusti commented:
Just got back from vacation and saw this. Really glad to hear you're gonna dig into this, @nhunzaker! I'll try to update more thoroughly once I'm settled. However, I actually found that the cross-iframe React input selection restoration completely broke focus and blur functionality for all inputs with the release of Chrome v60, so I updated our React form build to completely remove the plugin, and it'd been working great for us since then. I have the suspicion that all of that code may actually have been in place for older browsers and/or React DOM rendering strategies that are no longer relevant or in use. However, making that determination would require knowing the exact set of circumstances that at some point were affected by that plugin, and there is not enough info in the comments for me to be sure of how to verify whether or not the plugin is still needed.

However, in light of our success with our app in production without any of that code, I've been planning to create yet another branch that completely removed that particular plugin and includes other changes necessary to make React octane-agnostic, and I will run yarn prettier on that branch, thanks for the suggestion!

Would love to hear what you're digging around uncovers about how to create the proper scenario for verifying the React input selection restoration functionality! Also, we can definitely close #7866 and #7936.

acusti commented:
@nhunzaker I’m keeping this open for now in case it turns out that the React input selection restoration plugin is still necessary with React fiber and latest browsers and such. However, the latest version of this effort that I think should be reviewed and tested is #11410.

nhunzaker commented:
Cool! Thanks for... sending out yet another PR. I really appreciate it.

acusti commented:
@nhunzaker I’ve pushed up a couple of commits that resolves the previously mentioned bug with this PR breaking focus and blur on inputs.

The first commit, d777546, adds an early return in `restoreSelection` that fixes the focus / blur bug and also renames some variables for clarity and accuracy.

The second commit, 7c59a77, is entirely optional. It just adds another early return as a performance optimization for the most common case of a single element with selection. I will leave it to others’ discretion to determine if it is worth leaving in, but I’d be happy to remove that commit.

acusti commented:
For anyone who wants to test these changes out in their own project, you can do so by updating the `react` and `react-dom` dependencies in your `package.json` to:

```json
    "react": "brandcast/react-built#a384c20",
    "react-dom": "brandcast/react-dom-built#3804391",
```

If you’re also using draft.js and want the full latest and greatest cross-iframe experience, use:

```json
    "draft-js": "brandcast/draft-js-built#003f24b",
```

@aweary Thanks again for the demonstration in the thread of #11410. Is it easy enough to update the `iframes-16` version to point to the build from this branch (`brandcast/react/iframes`)?

Also, I’d like to make a version of the demo that renders into multiple iframes (using `react-frame-component` or that approach) each with their own input, then make selections in each input in each iframe and move them around to see if they maintain visible selection. I updated my existing demo of one of the use cases for this feature (rendering across iframes) is useful to use the latest built versions of react, react-dom, and draft-js: https://codepen.io/acusti/pen/bWaJXe

acusti commented:
@nhunzaker In order to make code review easier, I thought I would annotate some of the major portions of the new code. Note that most of the 363 new lines of code are from the 254 lines in the new ReactInputSelection test file.

https://github.com/facebook/react/pull/9184/files#diff-a654f37b01573fc8006b426d56ad53ceR23

```js
getActiveElementDeep: () => HTMLElement;
```
The new `getActiveElementDeep` function gets the one true focused element in the DOM across any accessible (same domain) iframes, else the `HTMLBodyElement` for the active `window` if there is no other `document.activeElement` (matching the behavior of fbjs’s `getActiveElement`). This refers to the one `HTMLElement` that will receive keyboard input / get `:focus` styles from the CSS pseudo selector / have a blinking cursor, and there can only be one in the DOM, but it could be in any `window`, parent or nested.

https://github.com/facebook/react/pull/9184/files#diff-a654f37b01573fc8006b426d56ad53ceR37

```js
type ElementSelection = {
  element: HTMLElement,
  selectionRange: {start: number, end: number},
};
getElementsWithSelections: (acc?: Array<ElementSelection>, win?: window) => Array<ElementSelection>
```

The objects returned by this function match what was previously returned by `getSelectionInformation()`. The function recurses over every frame’s `window` object on the page, starting from the parent `window` object, and looking for any elements with existing visible selections, even if the element that has a selection isn’t the single `activeElement` in the state of the DOM that gets returned by the new `getActiveElementDeep`. Those elements should still generally still have a visible selection, though it will appear greyed out (inactive / not focused). Those elements get saved as `elementSelections` when `ReactInputSelection.getSelectionInformation` is invoked, along with the one true focused / active element, so that on restoreSelection, we can iterate over every element that had a visible selection and restore that selection if needed.

https://github.com/facebook/react/pull/9184/files#diff-a654f37b01573fc8006b426d56ad53ceR126

```js
type SelectionInformation = {
  activeElement: HTMLElement,
  elementSelections: Array<ElementSelection>,
};

ReactInputSelection.getSelectionInformation: () => SelectionInformation;
```

Previously this function returned ElementSelection for a single `document.activeElement`, with an optional `selectionRange` if the active element `hasSelectionCapabilities`. In order to be able to restore multiple element selections across multiple `window`s, however, I updated it to return an array of `ElementSelection`s, along with the one true `activeElement` element, because both are needed by `restoreSelection`. Speaking of:

https://github.com/facebook/react/pull/9184/files#diff-a654f37b01573fc8006b426d56ad53ceR138

```js
ReactInputSelection.restoreSelection: (SelectionInformation) => void;
```

This function has been updated to restore selection for 0 or more elementSelections along with the one true activeElement. The array of visible selections and the activeElement are both necessary because restoring the selection to the element that previously had a visible selection is not enough to make that element visible; it also needs to be focused by invoking `element.focus()`. However, this will make that element the real one true active element in the DOM, so to prevent the previous one true active element from being changed, at the end of restoring selection to each element that had a visible selection, we refocus the previously identified one true active element.

For more background about how visible selections work across multiple browsing contexts, see https://stackoverflow.com/a/40104212/333625 (and [this demo](https://codepen.io/acusti/pen/PGakzY?editors=0010)).

https://github.com/facebook/react/pull/9184/files#diff-a654f37b01573fc8006b426d56ad53ceR86

```js
focusNodePreservingScroll: (HTMLElement) => void;
```

I moved the code that was previously defined directly in `ReactInputSelection.restoreSelection` into a util function so that it can be invoked from multiple places.

nhunzaker commented on packages/react-dom/src/client/ReactDOMSelection.js:
```diff
@@ -153,11 +157,13 @@ export function getModernOffsetsFromPoints(
  */
 export function setOffsets(node, offsets) {
-  if (!window.getSelection) {
+  var doc = node.ownerDocument || document;
```
When is ownerDocument ever undefined?

acusti replied:
According to the MDN page about the `Node` interface:

>  `Node.ownerDocument` returns the `Document` that this node belongs to. If no document is associated with it, returns `null`.

That doesn’t refer to a node that isn’t in the DOM, which will still have a reference to the `ownerDocument` where it originated (via `document.createElement`). In my experience, `ownerDocument` is only `null` for `Document` nodes themselves, which matches what MDN says on the page specifically about `Node.ownerDocument`:

> If this property is used on a node that is itself a document, the result is `null`.

So now that you mention it, this check could be updated to handle `Document` specifically, maybe like `var doc = node.nodeType === 9 ? node : node.ownerDocument;`

Or it could just continue to fallback to the global document whenever `ownerDocument` isn’t truthy.

nhunzaker replied:
Cool. Thank you. That was extremely helpful!

nhunzaker commented on packages/react-dom/src/client/ReactInputSelection.js:
```diff
@@ -13,7 +13,95 @@ import * as ReactDOMSelection from './ReactDOMSelection';
+      win = element.contentDocument.defaultView;
+    } catch (e) {
+      return element;
+    }
```
Why a try/catch? Could you just do:

```
if (!element.contentDocument) {
  return element
}
```
?

acusti replied:
If you just try to access the `contentDocument` of an `HTMLIframeElement`that comes from a non-marching separate domain (like if there’s an embedded YouTube player in the DOM), the browser will throw an error.

nhunzaker commented:
Took a quick review pass, but quickly discovered that I have some background to catch up on. Digging in.

acusti commented:
@nhunzaker Excellent! Please don’t hesitate to ask me questions if you have anything in particular you want to know more about. It took me quite a bit of research(a to figure out how this all should and does work, and I would be delighted to pass some of that on where helpful. And at this point, it is a major business priority for us to get this PR correct and merged, so I am more than happy to make myself available as a resource in any capacity that could be useful.

acusti commented:
@nhunzaker Just merged latest `master` (v16.1.1) and resolved all conflicts. Also ran `yarn prettier-all` again and committed the results, which I’m hoping will fix the build. Anything else I can do to help move this along? Thanks as always.

nhunzaker commented:
@acusti I've blocked out time later this afternoon to look at this PR. I am sorry it has taken this long. Thank you for all of your work!

nhunzaker commented on packages/react-dom/src/__tests__/ReactInputSelection-test.js:
```diff
@@ -0,0 +1,254 @@
+      expect(ReactInputSelection.hasSelectionCapabilities(inputHidden)).toBe(
+        false,
+      );
+    });
```
These feel like they could be separate tests. 

Also: Why do `email`, `password`, and `number` return false? Is this because they don't support the text selection api?

acusti replied:
That’s a great question! That’s what the existing function enforces, so I figured I’d capture that behavior in the tests:

```js
  return (
    nodeName &&
    ((nodeName === 'input' && elem.type === 'text') ||
      nodeName === 'textarea' ||
      elem.contentEditable === 'true')
  );
```

But I’m not sure it’s desired, because those input types can definitely have selections. The `hasSelectionCapabilities` check could be changed to something like:
```js
  return (
    nodeName &&
    ((nodeName === 'input' && ['text', 'email', 'number', 'password'].includes(elem.type)) ||
      nodeName === 'textarea' ||
      elem.contentEditable === 'true')
  );
```

nhunzaker commented on packages/react-dom/src/__tests__/ReactInputSelection-test.js:
```diff
@@ -0,0 +1,254 @@
+      var body = document.createElement('body');
+      body.contentEditable = 'true';
+      var input = document.createElement('input');
+      input.contentEditable = 'true';
```
Huh. I didn't know you could do this.

nhunzaker replied:
<img width="438" alt="screen shot 2017-11-16 at 4 39 22 pm" src="https://user-images.githubusercontent.com/590904/32917297-ba11e4f4-caec-11e7-8e0b-cf010973cb5b.png">

How fun.

nhunzaker commented on packages/react-dom/src/client/ReactInputSelection.js:
```diff
@@ -49,34 +137,38 @@ export function getSelectionInformation() {
-    }
+  if (
+    curActiveElement !== priorActiveElement &&
+    isInDocument(priorActiveElement)
```
Possible to recycle this value from the check on line 147?

acusti replied:
The check is actually useless, because if `priorActiveElement` is not in document, the function will early return and never reach this if block.

nhunzaker commented on packages/react-dom/src/events/SelectEventPlugin.js:
```diff
@@ -64,32 +64,44 @@ function getSelection(node) {
+  var doc =
+    nativeEventTarget.ownerDocument ||
+    nativeEventTarget.document ||
+    nativeEventTarget;
```
Could you explain what `nativeEventTarget` can be? When would an element be missing an owner document, but have a document? What are the usual values of`nativeEventTarget`?

acusti replied:
If the event was dispatched at the `window` level, like if you had a `window.onresize` listener, for example, the event target would be the `window` object and would therefore have a `document` but no `ownerDocument`. And if it was dispatched on the `window` object of a nested browsing context, it would be the `window` object of that iframe, not the global one.

However, looking at where this gets invoked, I see that the call site has it’s own nested browsing context-compatible way to calculate `doc` that was added a year ago: https://github.com/facebook/react/commit/6e04bd758ed7a28ebdfbf544f6db9065340392e1

I guess I’ll just repeat the same logic here, if that makes sense to you; i.e. I’ll replace these lines with:

```js
    var doc =
      nativeEventTarget.window === nativeEventTarget
        ? nativeEventTarget.document
        : nativeEventTarget.nodeType === DOCUMENT_NODE
          ? nativeEventTarget
          : nativeEventTarget.ownerDocument;
```

acusti replied:
Considering they’re in the same file, I’m making a `getEventTargetDocument` function. I dropped `native` from the name because I don’t think that distinction is relevant at the level of this function.

nhunzaker commented on packages/react-dom/src/events/SyntheticClipboardEvent.js:
```diff
@@ -13,9 +13,11 @@ import SyntheticEvent from 'events/SyntheticEvent';
+    var doc = (event.target && event.target.ownerDocument) || document;
+    return doc.defaultView.clipboardData;
   },
 };
```
What does clipboard data have to do with selection?

acusti replied:
Nothing to do with selection; this is just another part of the codebase that relies on the global `window` object rather than the `window` object relative to the DOM node being used. That said, I just looked it up, and `window.clipboardData` is an IE-only API and one where it seems like the code being executed should only try to look up it’s own `window.clipboardData` (because of browser permissions / security settings), so I’m going to revert this set of changes.

nhunzaker replied:
Ah still, do you think it's worth landing? What do you think about sending this out in a separate PR?

acusti replied:
I don’t know really know anything about clipboard APIs in IE, but based on a quick google search, my hunch is that this change wouldn’t actually be for the better. There’s some mentions about browser settings and issues with clipboardData not being accessible across different browsing contexts, so keeping it hardcoded to `window.clipboardData` might actually be correct.

But I’d be happy to open a separate PR with it if someone would be able to figure that all out.

acusti commented:
@nhunzaker Thanks so much for the review! I addressed all the comments and pushed up the changes. I also pushed up a fix for a bug in `getElementsWithSelections` which could lead to an element that is a document’s `activeElement` losing focus from `restoreSelection` trying to focus another element in the same document: https://github.com/facebook/react/pull/9184/commits/174375cf24cda80d3fe7da3573472d015434c870

The bug came up in our internal team’s QA of this branch of work in our app and is fixed by that change.

I’d also be happy to update `hasSelectionCapabilities` to support other input types than just `text`, if that makes sense to you. And depending on if I should do that or not, I will update the tests.

acusti commented:
@nhunzaker I decided to go ahead and update `hasSelectionCapabilities` and its tests to be correct in its return value for `<input>`s, no point in preserving the current incorrect behavior. I also broke the single test into two tests, one for input types that return true, one for those that return false. The commit: https://github.com/facebook/react/pull/9184/commits/82443f7738205ec6ee29bc1f5701089a9c4fadf2

acusti commented:
@nhunzaker I merged latest master and resolved the conflicts. Also added a guard based on a test failure and converted all `var` to make eslint pass. As of your last comment, you mentioned wanting to create a manual DOM fixture, which sounds like a good idea. I took a quick look and wasn’t sure where to start, though I definitely could give you some pretty stripped down code examples that can be used for those tests. For example, in [this codepen demo](https://codepen.io/acusti/pen/bWaJXe?editors=0010), I use the following component which renders its children into an iframe:

```js
class Iframe extends React.Component {
	componentDidMount() {
		const doc = this.getDocument();
		doc.open();
		doc.write(`<!doctype html><html><head><style>${iframeStyles}</style><body><div></div></body></html>`);
		doc.close();

		this.renderIframeContents();
	}

	componentDidUpdate() {
		this.renderIframeContents();
	}

	getDocument() {
		return ReactDOM.findDOMNode(this).contentWindow.document;
	}

	renderIframeContents() {
		const doc = this.getDocument();
		const mountTarget = doc.body.children[0];
		ReactDOM.unstable_renderSubtreeIntoContainer(this, this.props.children, mountTarget);
	}

	render() {
		return <iframe />;
	}
}
```

That component could be combined with an example like the one @aweary posted in another PR thread to create a fixture that verifies that all of the `restoreSelection` functionality continues to work across nested browsing contexts: http://react-master-restore-selection.surge.sh/

You also mentioned wanting to ping Dan when he was back from vacation. Thinking you all might be back on vacation again, but maybe 2018 will be a good opportunity to ping him and get this rolling again.

gaearon commented:
What is the difference in bundle size?

Maybe it can be okay if:

* the bundle size doesn't increase
* the unit tests stop relying on private APIs and test only public APIs instead
* somebody on the team can shepherd this through

But I'm not sure this case is important enough to offset the complexity it introduces, and I'm not very optimistic about this landing. I really appreciate your effort though.

ahutchings commented:
I would love for this change to make it in. Using React portals greatly simplifies multi-window apps using the [pattern documented here](https://hackernoon.com/using-a-react-16-portal-to-do-something-cool-2a2d627b0202).

There does seem to be a fair amount of demand for rendering into iframes as well - [react-frame-component](https://www.npmjs.com/package/react-frame-component) has 39 dependent packages and 20,627 downloads over the past month.

In my opinion, the complexity required in fixing multi-context selection seems a worthwhile trade off in order to fully support this class of interesting React usage patterns.

gaearon commented:
Fair enough. I'd like to see the bundle size difference I guess,

acusti commented:
@gaearon For the bundle size difference, I did a fresh `yarn build` and committed `scripts/rollup/results.json` in 96ecdf744c6372dc987b2bfb1e5cff97de340a58. Diff is reproduced below:

```diff
From 96ecdf744c6372dc987b2bfb1e5cff97de340a58 Mon Sep 17 00:00:00 2001
From: Andrew Patton <andrew@acusti.ca>
Date: Sat, 6 Jan 2018 20:30:57 -0800
Subject: [PATCH 19/19] Record sizes

---
 scripts/rollup/results.json | 24 ++++++++++++------------
 1 file changed, 12 insertions(+), 12 deletions(-)

diff --git a/scripts/rollup/results.json b/scripts/rollup/results.json
index 882989be0f0..17de7e10ae2 100644
--- a/scripts/rollup/results.json
+++ b/scripts/rollup/results.json
@@ -46,43 +46,43 @@
       "filename": "react-dom.development.js",
       "bundleType": "UMD_DEV",
       "packageName": "react-dom",
-      "size": 560498,
-      "gzip": 132755
+      "size": 563490,
+      "gzip": 133475
     },
     {
       "filename": "react-dom.production.min.js",
       "bundleType": "UMD_PROD",
       "packageName": "react-dom",
-      "size": 93742,
-      "gzip": 30770
+      "size": 95062,
+      "gzip": 31234
     },
     {
       "filename": "react-dom.development.js",
       "bundleType": "NODE_DEV",
       "packageName": "react-dom",
-      "size": 544531,
-      "gzip": 128982
+      "size": 547519,
+      "gzip": 129732
     },
     {
       "filename": "react-dom.production.min.js",
       "bundleType": "NODE_PROD",
       "packageName": "react-dom",
-      "size": 92163,
-      "gzip": 29827
+      "size": 93483,
+      "gzip": 30255
     },
     {
       "filename": "ReactDOM-dev.js",
       "bundleType": "FB_DEV",
       "packageName": "react-dom",
-      "size": 562017,
-      "gzip": 131222
+      "size": 565082,
+      "gzip": 132016
     },
     {
       "filename": "ReactDOM-prod.js",
       "bundleType": "FB_PROD",
       "packageName": "react-dom",
-      "size": 265467,
-      "gzip": 51292
+      "size": 266674,
+      "gzip": 51845
     },
     {
       "filename": "react-dom-test-utils.development.js",
```
------------
Being able to render into an iframe is very powerful. It’s actually the only way to properly solve two problems:

- Enabling multiple visible selections in the same app at once: many apps or text editors fake this using a bunch of special selection state management, spans, and background colors, but it is a degraded experience (guaranteed not to match the native behavior of all OSes where it is run), but you can get the native browser / OS experience by having the different selectable elements in separate browsing contexts
- 100% reliable CSS encapsulation (e.g. render a preview of something with custom CSS without worrying about loading those styles into the app window): in our case, we are building a webpage design-focused builder, so using an iframe means we can even change the dimensions of the iframe in order to preview how the webpage will look at different device sizes (mobile, tablet, etc)

If you want a demo of both of those use cases in action (using this fork of react plus the draft-js fork) and a pretty compelling case for why these changes are worth it, check out [Brandcast](https://app.brandcast.io/signup/), our app. We recently replaced our usage of `react-frame-component` since the release of React 16 with a component that does the same thing via the Portal API (instead of `ReactDOM.unstable_renderSubtreeIntoContainer`).

In terms of complexity of this PR, I could simplify the changes necessary to make React work across nested browsing contexts (iframes or separate windows in multi-window apps). This set of changes is the most complete solution that handles any number of elements with visible selections across any number of nested browsing contexts with any level of recursion. I made another PR that takes the opposite approach and removes the React input selection restoration plugin entirely: #11410. However, as @sophiebits and @aweary [pointed](https://github.com/facebook/react/pull/11410#issuecomment-340847837) [out](https://github.com/facebook/react/pull/11410#issuecomment-340847931), the plugin does still serve a purpose, though it’s a pretty edge-casey one.

My first attempt at this set of changes, #7866 (from October 2016), implemented yet another different approach that was simpler but less complete. It only tries to preserve and restore the one active selection in the app and does nothing about any other visible (but inactive) selections. I would be willing to redo that approach against latest `master` (things have changed over the past year and a quarter) if that felt to the React core team like a better overall approach, and it would result in a smaller increase in the size of the bundle.

---------------

Lastly, @gaearon, you mentioned:
> the unit tests stop relying on private APIs and test only public APIs instead

Are you referring to the exported `ReactInputSelection.hasSelectionCapabilities` and `ReactInputSelection.getSelection` methods? I’d be happy to remove those tests; I just added them because I figured it would be a simpler, more maintainable, more pure way to test the functionality of that plugin. Would you prefer to have tests that only interface with `ReactInputSelection.getSelectionInformation` and `ReactInputSelection.restoreSelection`, the two functions that get used by `ReactDOM`? Or does that still qualify as a private API? Or am I misunderstanding your suggestion entirely?

Thanks for checking out the PR and giving your feedback! And as I mentioned previously to @nhunzaker, don’t hesitate to ask me for anything, whether info, documentation, questions, code, etc.

flarnie commented:
I'm looking at the related PR on Draft.js that would enable iframe support there, and it would be helpful to get a verdict on this before I consider that one.

I'm going to bring this to the other maintainers and hopefully we can get an answer one way or another. Thanks for your ongoing work on this @acusti and to all the folks guiding this work.

flarnie commented:
Just an update - the React team at Facebook has discussed this a bit, still need to discuss more, but it is on our radar and hope to have an answer soon.

acusti commented:
@flarnie @gaearon I decided to go ahead and create a new PR that implements the minimal changes required to make React work across nested browsing contexts (with 0% or even -1% changes to the bundle size), as referred to in my [last comment](https://github.com/facebook/react/pull/9184#issuecomment-355800102). The PR: #12037

I hope between these two approaches we can find something that everyone feels comfortable with being merged.

flarnie commented:
:) Thanks for doing that @acusti - we are looking for a maintainer to take point on this; might take a while but it is not forgotten about.
